crates/ty_python_semantic/src/types/generics.rs

```diff
@@ -430,6 +432,115 @@ pub(super) fn walk_specialization<'db, V: super::visitor::TypeVisitor<'db> + ?Si
     }
 }
 
+fn is_subtype_in_invariant_position<'db, C: Constraints<'db>>(
+    db: &'db dyn Db,
+    derived_type: &Type<'db>,
+    derived_materialization: MaterializationType,
+    base_type: &Type<'db>,
+    base_materialization: MaterializationType,
+) -> C {
+    let derived_top = derived_type.top_materialization(db);
+    let derived_bottom = derived_type.bottom_materialization(db);
+    let base_top = base_type.top_materialization(db);
+    let base_bottom = base_type.bottom_materialization(db);
+    match (derived_materialization, base_materialization) {
+        // Top and bottom materializations are fully static, so subtyping and assignability are the same.
+        // One top is a subtype of another if the range of materializations covered by derived
+        // is a subset of the range covered by base.
+        // This is also true for subtyping between None and Top.
+        // And assignability between Top and None.
+        (MaterializationType::Top, MaterializationType::Top) => C::from_bool(
+            db,
+            base_bottom.is_subtype_of(db, derived_bottom)
+                && derived_top.is_subtype_of(db, base_top),
+        ),
+        // One bottom is a subtype of another if it covers a strictly larger set of materializations.
+        (MaterializationType::Bottom, MaterializationType::Bottom) => C::from_bool(
+            db,
+            derived_bottom.is_subtype_of(db, base_bottom)
+                && base_top.is_subtype_of(db, derived_top),
+        ),
+        // A bottom materialization is a subtype of top if it's a subtype of some part of the range of materializations
+        // covered by top, i.e. either the top or bottom of derived is between the top and derived of base, or both
+        // materializations of base are between the top and bottom of derived.
```

Comment:
I think the referent of "it's" in the first phrase here is not clear. It is not the case that "the bottom materialization" mentioned previously must be "a subtype of some part of the range of materializations covered by top." (That would be a simpler implementation, which uses only `derived_bottom` and not `derived_top`.)

I think what is implemented here (and described after the "i.e.") is "some materialization of derived must be a subtype of some materialization of base."

Also, "top and derived of base" should be "top and bottom of base".

```suggestion
        // The bottom materialization of `Derived` is a subtype of the top materialization of `Base` if some
        // materialization of `Derived` is a subtype of some materialization of `Base`, i.e. either the top or
        // bottom of `Derived` is between the top and bottom of `Base`, or both materialization of `Base`
        // are between the top and bottom of `Derived`.
```